src/block-data-providers/bitcoin-core/provider.ts

```diff
@@ -0,0 +1,278 @@
+        this.config = this.configService.get<BitcoinCoreConfig>('bitcoinCore');
+    }
+
+    async onApplicationBootstrap() {
+        const getState = await this.getState();
+        if (getState) {
+            this.logger.log(
+                `Restoring state from previous run: ${JSON.stringify(
+                    getState,
+                )}`,
+            );
+        } else {
+            this.logger.log('No previous state found. Starting from scratch.');
+            const state: BitcoinCoreOperationState = {
+                currentBlockHeight: 0,
+                indexedBlockHeight:
+                    this.configService.get<BitcoinNetwork>('app.network') ===
+                    BitcoinNetwork.MAINNET
+                        ? TAPROOT_ACTIVATION_HEIGHT - 1
+                        : 0,
+            };
+            await this.setState(state);
+        }
+    }
+
+    @Cron(CronExpression.EVERY_10_SECONDS)
+    async sync() {
+        if (this.isSyncing) return;
+        this.isSyncing = true;
+
+        const state = await this.getState();
+        if (!state) {
+            throw new Error('State not found');
+        }
+        const tipHeight = await this.getTipHeight();
+        if (tipHeight <= state.indexedBlockHeight) {
+            this.logger.debug(
+                `No new blocks found. Current tip height: ${tipHeight}`,
+            );
+            this.isSyncing = false;
+            return;
+        }
+
+        let height = state.indexedBlockHeight + 1;
+
+        for (height; height <= tipHeight; height++) {
+            const transactions = await this.processBlock(height);
+            for (const transaction of transactions) {
+                const { txid, vin, vout, blockHeight, blockHash } = transaction;
+                await this.indexTransaction(
+                    txid,
+                    vin,
+                    vout,
+                    blockHeight,
+                    blockHash,
+                );
+            }
+            state.indexedBlockHeight = height;
+            await this.setState(state);
+        }
+        this.isSyncing = false;
+    }
+
+    async getTipHeight(): Promise<number> {
+        try {
+            const body = {
+                method: 'getblockcount',
+                params: [],
+                ...this.createPartialRPCBody(),
+            };
+            return await this.request(body);
+        } catch (error) {
+            this.logger.log(`Error fetching block count`);
+            throw error;
+        }
+    }
+
+    async getBlockHash(height: number): Promise<string> {
+        try {
+            const body = {
+                method: 'getblockhash',
+                params: [height],
+                ...this.createPartialRPCBody(),
+            };
+            return await this.request(body);
+        } catch (error) {
+            this.logger.log(`Error fetching  block hash of height : ${height}`);
+            throw error;
+        }
```

Comment:
Let's remove try-catch from here and move it inside `this.request()`. We'll probably end up moving it there once we implement retry logic so it's better to do it now.

Reply:
Also, log the request body and response so that we can know what went wrong.

Reply:
I see that you are already logging inside `this.request` so I don't get why we are doing it again here.